sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ParDoTest.java

```diff
@@ -3808,15 +3817,24 @@ public void onTimer(
       TestStream<KV<String, Integer>> stream =
           TestStream.create(KvCoder.of(StringUtf8Coder.of(), VarIntCoder.of()))
               // See GlobalWindow,
-              // END_OF_GLOBAL_WINDOW is TIMESTAMP_MAX_VALUE.minus(Duration.standardDays(1))
-              .advanceWatermarkTo(BoundedWindow.TIMESTAMP_MAX_VALUE.minus(Duration.standardDays(1)))
+              // END_OF_GLOBAL_WINDOW is TIMESTAMP_MAX_VALUE.minus(Duration.standardDays(1)),
```

Comment:
Updating other tests to second granularity makes sense but changing the test to see how END_OF_GLOBAL_WINDOW is interpreted changes what this test is explicitly trying to do.

Reply:
My understanding is that it's trying to test setting a timer with timestamp that exceed END_OF_GLOBAL_WINDOW and the fire timestamp gets altered to the END_OF_GLOBAL_WINDOW(gc time). How the original exceeding-limit timestamp is obtained shouldn't make a difference?
Since END_OF_GLOBAL_WINDOW is with millis granularity, advance the watermark of the TestStream to this time will not be processed by dataflow.

Reply:
My issue is that this changes the test such that the boundary condition is looser by making END_OF_GLOBAL_WINDOW looser.

@acrites Any ideas/suggestions here?

Reply:
we can also explicitly set timer with absolute time TIMESTAMP_MAX_VALUE, which set the timer fire time the same to original `timer.align(Duration.standardDays(1)).setRelative();`, would that make sense?

Reply:
Yichi and I played around with some flag settings in Dataflow and I think we can change the time resolution for tests to milliseconds, which should make this test pass. However, I'm a little worried in that we then aren't testing the same as prod, so we might end up missing some other bugs that come up.

Otherwise, I don't really know a good way around this since `BoundedWindow.TIMESTAMP_MAX_VALUE` is not a round number of seconds.

Reply:
@acrites and I chatted about some of the tests. There is an option to alter the windmill wm_work_set_watermark_resolution_usec and dataflow_watermark_resolution_usec so that the AdvanceWatermark and AdvanceProcessingTime can deal with milliseconds. But it requires additional test framework plumbing. Should we keep this test as it was and try to unsickbay it after we can set the windmill flag values?

Reply:
Lets leave this one tests as is and remove it from the change. We can update the others with the changes since they won't change the semantics of what is being tested.

For Dataflow's TestStream implementation, to simulate production, it should round all the advance timestampts to the closest `>=` internal timestamp. It should also make sure that no two advance timestamps that aren't the same are not rounded to the same value. I believe this would make this test work and it would simulate what is happening in production.